src/labone/nodetree/helper.py

```diff
@@ -0,0 +1,269 @@
+
+    """
+    current_dict: NestedDict[T] | T = nested_dict
+    try:
+        for key in key_chain:
+            current_dict = current_dict[key]  # type: ignore[index]
+    except TypeError as e:
+        msg = (
+            f"{nested_dict} cannot be indexed with {key_chain}, because key-sequence"
+            f" is to long and hits a leaf early"
+        )
+        raise KeyError(msg) from e
+    return current_dict
+
+
+def join_path(*args: NormalizedPathSegment) -> LabOneNodePath:
+    """Join path in a well-defined manner.
+
+    Node that a leading seperator is always added.
+
+    Args:
+        args: Segments to join.
+
+    Returns:
+        One joint path.
+    """
+    return PATH_SEPERATOR + PATH_SEPERATOR.join(args)
+
+
+def split_path(path: LabOneNodePath) -> list[NormalizedPathSegment]:
+    """Split path in a well-defined manner.
+
+    A leading seperator is ignored.
+
+    Args:
+        path: Path to be split.
+
+    Returns:
+        Segments of which the path consists.
+    """
+    path_segments = path.split(PATH_SEPERATOR)
+    first_item_index = 0
+    if path_segments[0] == "":  #
+        # this happens if the path started with '/'
+        # ignore leading '/'
+        first_item_index = 1
+    return path_segments[first_item_index:]
+
+
+def get_prefix(path: LabOneNodePath, segment_count: int) -> LabOneNodePath:
+    """Get a path consisting only of the first n segments of the original path.
+
+    Args:
+        path: Path, the prefix should be taken from.
+        segment_count: Number of segments, which should be included. If there are
+            not enough segments, all available ones are included.
+
+    Returns:
+        A path, similar to the given one, but bounded in terms of number of segments.
+    """
+    segments = split_path(path)
+    first_segments = segments[:segment_count]
+    return join_path(*first_segments)
+
+
+def normalize_path_segment(path_segment: str | int) -> NormalizedPathSegment:
+    """Bring segment into a standard form.
+
+    - no integers, but only strings
+    - '_' in reserved names ignored
+
+    Args:
+        path_segment: Segment of a path to be normalized.
+
+    Returns:
+        The segment, following the described formatting standards.
+
+    """
+    return str(path_segment).lower().rstrip("_")
+
+
+def pythonify_path_segment(path_segment: NormalizedPathSegment) -> str:
+    """Try to bring segment into a form, which can be used as an attribute for a node.
+
+    - add '_' at end of reserved names
+    - make clear that numbers should be used with indexing e.g. [0] instead of node.0
+
+    Args:
+        path_segment: Segment in the usual representation.
+
+    Returns:
+        Path segment in pythonic representation.
+
+    """
+    try:
+        int(path_segment)
+    except ValueError:
+        pass
+    else:
+        return (
+            f"(you can use indexing like: [{path_segment}] "
+            f"and remove this description)"
+        )
```

Comment:
since there is no way to use a number as an attribute, we are in a kind of dilemma. I tried hard to avoid showing plain numbers, because node.0 is invalid and will confuse is suggested. The current solution is certainly worse, but what about [0] ? I think an autocompleted node.[0] is better invalid than node.0 as it shows the solution right away

Reply:
All of them are ugly thats why I would go with the most simple. Do what toolkit does and just add the plain number to the __dir__ method.